pkg/storage/unified/entitybridge/entitybridge.go

```diff
@@ -0,0 +1,271 @@
+		Guid:         string(event.Object.GetUID()),
+		GroupVersion: gvr.Version,
+
+		Folder:  obj.GetFolder(),
+		Body:    event.Value,
+		Message: event.Object.GetMessage(),
+
+		Labels: obj.GetLabels(),
+		Size:   int64(len(event.Value)),
+	}
+
+	switch event.Type {
+	case resource.WatchEvent_ADDED:
+		msg.Action = entity.Entity_CREATED
+		rsp, err := b.client.Create(ctx, &entity.CreateEntityRequest{Entity: msg})
+		if err != nil {
+			return 0, err
+		}
+		return rsp.Entity.ResourceVersion, err
+
+	case resource.WatchEvent_MODIFIED:
+		msg.Action = entity.Entity_UPDATED
+		rsp, err := b.client.Update(ctx, &entity.UpdateEntityRequest{
+			Entity:          msg,
+			PreviousVersion: event.PreviousRV,
+		})
+		if err != nil {
+			return 0, err
+		}
+		return rsp.Entity.ResourceVersion, err
+
+	default:
+	}
+
+	return 0, fmt.Errorf("unsupported operation: %s", event.Type.String())
+}
+
+func (b *entityBridge) WatchWriteEvents(ctx context.Context) (<-chan *resource.WrittenEvent, error) {
+	client, err := b.client.Watch(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &entity.EntityWatchRequest{
+		Action:            entity.EntityWatchRequest_START,
+		Labels:            map[string]string{},
+		WithBody:          true,
+		WithStatus:        true,
+		SendInitialEvents: false,
+	}
+
+	err = client.Send(req)
+	if err != nil {
+		err2 := client.CloseSend()
+		if err2 != nil {
+			klog.Errorf("watch close failed: %s\n", err2)
+		}
+		return nil, err
+	}
+
+	reader := &decoder{client}
+	stream := make(chan *resource.WrittenEvent, 10)
+	go func() {
+		for {
+			evt, err := reader.next()
+			if err != nil {
+				reader.close()
+				close(stream)
+				return
+			}
+			stream <- evt
+		}
+	}()
+	return stream, nil
+}
+
+// IsHealthy implements ResourceServer.
+func (b *entityBridge) IsHealthy(ctx context.Context, req *resource.HealthCheckRequest) (*resource.HealthCheckResponse, error) {
+	rsp, err := b.client.IsHealthy(ctx, &entity.HealthCheckRequest{
+		Service: req.Service, // ??
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &resource.HealthCheckResponse{
+		Status: resource.HealthCheckResponse_ServingStatus(rsp.Status),
+	}, nil
+}
+
+// Read implements ResourceServer.
+func (b *entityBridge) Read(ctx context.Context, req *resource.ReadRequest) (*resource.ReadResponse, error) {
+	v, err := b.client.Read(ctx, &entity.ReadEntityRequest{
+		Key:      toEntityKey(req.Key),
+		WithBody: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &resource.ReadResponse{
+		ResourceVersion: v.ResourceVersion,
+		Value:           v.Body,
+	}, nil
+}
+
+// List implements ResourceServer.
+func (b *entityBridge) PrepareList(ctx context.Context, req *resource.ListRequest) (*resource.ListResponse, error) {
+	key := req.Options.Key
+	query := &entity.EntityListRequest{
+		NextPageToken: req.NextPageToken,
+		Limit:         req.Limit,
+		Key:           []string{toEntityKey(key)},
+		WithBody:      true,
+	}
+
+	// Assumes everything is equals
```

Comment:
we should just pass stuff that is equals, and handle other predicates in storage